### PR TITLE
network: slcan support

### LINKED ISSUE #13150: networkd: interfaces using slcan driver are not recognized as CAN devices
**systemd version the issue has been seen with**
> 241

**Used distribution**
 > Yocto/Poky

**Expected behaviour you didn't see**
 > `slcan` intefaces should be configured as CAN devices (ie. don't try to add IP addresses)

**Unexpected behaviour you saw**
 > networkd does not recognize `slcan` interfaces as CAN devices and tries to configure them as normal network interfaces.

**Steps to reproduce the problem**
1. Attach a serial CAN interface to the system using `slcand`.
2. Create a `.network` file for the interface.
3. Observe that the interface fails to be brought up.

The direct cause of this issue is that the `slcan` driver does not support the `IFLA_INFO_KIND` netlink attribute, which networkd uses to identify CAN devices. One solution is to add support for that attribute to the `slcan` driver in the kernel, as shown below:
<details>
  <summary>slcan driver patch</summary>

```
From ee0c06f206df2e5de18688b5d3f0b126117b42c1 Mon Sep 17 00:00:00 2001
From: Ben Wolsieffer <bwolsieffer@automarinesys.com>
Date: Wed, 10 Jul 2019 16:36:53 -0400
Subject: [PATCH 8/8] can: slcan: support IFLA_INFO_KIND rtnl attribute

---
 drivers/net/can/slcan.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/net/can/slcan.c b/drivers/net/can/slcan.c
index aa97dbc797b6..ccdfc146baa7 100644
--- a/drivers/net/can/slcan.c
+++ b/drivers/net/can/slcan.c
@@ -55,6 +55,7 @@
 #include <linux/workqueue.h>
 #include <linux/can.h>
 #include <linux/can/skb.h>
+#include <net/rtnetlink.h>
 
 MODULE_ALIAS_LDISC(N_SLCAN);
 MODULE_DESCRIPTION("serial line CAN interface");
@@ -450,6 +451,11 @@ static void slc_setup(struct net_device *dev)
 	dev->features           = NETIF_F_HW_CSUM;
 }
 
+static struct rtnl_link_ops slc_link_ops __read_mostly = {
+	.kind	= "can",
+	.setup	= slc_setup,
+};
+
 /******************************************
   Routines looking at TTY side.
  ******************************************/
@@ -532,6 +538,7 @@ static struct slcan *slc_alloc(void)
 		return NULL;
 
 	dev->base_addr  = i;
+	dev->rtnl_link_ops = &slc_link_ops;
 	sl = netdev_priv(dev);
 
 	/* Initialize channel control data */
@@ -703,6 +710,10 @@ static int __init slcan_init(void)
 	pr_info("slcan: serial line CAN interface driver\n");
 	pr_info("slcan: %d dynamic interface channels.\n", maxdev);
 
+	status = rtnl_link_register(&slc_link_ops);
+	if (status)
+		return status;
+
 	slcan_devs = kcalloc(maxdev, sizeof(struct net_device *), GFP_KERNEL);
 	if (!slcan_devs)
 		return -ENOMEM;
@@ -769,6 +780,8 @@ static void __exit slcan_exit(void)
 	kfree(slcan_devs);
 	slcan_devs = NULL;
 
+	rtnl_link_unregister(&slc_link_ops);
+
 	i = tty_unregister_ldisc(N_SLCAN);
 	if (i)
 		printk(KERN_ERR "slcan: can't unregister ldisc (err %d)\n", i);
-- 
2.22.0
```

</details>

The problem with this approach is that the `IFLA_INFO_KIND` attribute seems to usually be set to the name of the driver rather than a string that unambiguously denotes that the link is a CAN device. If I were to submit this kernel patch upstream, I may be asked to change the attribute value to "slcan". This would require another special case in systemd ([like what is done for `vxcan` and `vcan`](https://github.com/systemd/systemd/blob/master/src/network/networkd-link.c#L2488)) to match this value. This check is done in many places and seems like a fragile way to detect a CAN device.

On the other hand, there seems to be a way to unambiguously detect a CAN device, using `net_device.type` (on the kernel side), which appears to be exposed in networkd as `Link.iftype`. All CAN drivers set its value to `ARPHRD_CAN`. Does this seem like a reasonable/better approach?

### PR DESCRIPTION
**Pull request by yuwata**
Closes #13150.

(Not tested.)

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -126,6 +129,9 @@ static bool link_dhcp4_enabled(Link *link) {
         if (link->network->bond)
                 return false;
 
+        if (link->iftype == ARPHRD_CAN)
+                return false;
+
         if (STRPTR_IN_SET(link->kind, "can", "vcan", "vxcan"))
                 return false;
```
Comment (poettering) [inline src/network/networkd-link.c:137]: maybe merge this into a new helper:

```c
static inline bool LINK_IS_CAN(Link *l) {
        assert(l);
        if (l->iftye == ARPHRD_CAN)
                return true;
        return STRPTR_IN_SET(link->kind, "can", "vcan", "vxcan");
}
```

**Comment by yuwata** [comment]
@poettering Thank you for the review. I've added the helper function. PTAL.

**Comment by yuwata** [comment]
CI failures are unrelated.

File: src/network/networkd-link.c (hunk 1 above)
Comment (lopsided98) [reply src/network/networkd-link.c:137]: Why is the `link->kind` check still necessary? AFAIK, every CAN driver has `iftype == ARPHRD_CAN`.

Comment (yuwata) [reply src/network/networkd-link.c:137]: Indeed. I will submit follow-up PR. Thanks.

Comment (yuwata) [reply src/network/networkd-link.c:137]: #13178.

**Pull request merged**